OneSignalSDK/onesignal/src/main/java/com/onesignal/FocusTimeController.java

```diff
@@ -0,0 +1,262 @@
+package com.onesignal;
+
+import android.os.SystemClock;
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+import android.support.annotation.WorkerThread;
+
+import org.json.JSONException;
+import org.json.JSONObject;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+
+/**
+ * Entry points that could case a state change to fire on_focus;
+ * 1. State change from OSSessionManager
+ * 2. App is foregrounded - Set start focused time
+ * 3. App is backgrounded - Kick off job (if needed)
+ *
+ * Complete set;
+ * 1. OneSignalSyncServiceUtils firing, cold starting the process
+ * 2. App cold starting via app open.
+ * 3. App cold starting via opening a notification.
+ * 4. App resuming via app open.
+ * 5. App resuming via a opening a notification.
+ */
+
+
+class FocusTimeController {
+   private long timeFocusedAtMs;
+
+   private static FocusTimeController sInstance;
+
+   private List<FocusTimeProcessorBase> focusTimeProcessors =
+      Arrays.asList(new FocusTimeProcessorUnattributed(), new FocusTimeProcessorAttributed());
+
+   private enum FocusEventType {
+      BACKGROUND,
+      END_SESSION
+   }
+
+   private FocusTimeController() { }
+   public static synchronized FocusTimeController getInstance() {
+      if (sInstance == null)
+         sInstance = new FocusTimeController();
+      return sInstance;
+   }
+
+   void appForegrounded() {
+      timeFocusedAtMs = SystemClock.elapsedRealtime();
+   }
+
+   void appBackgrounded() {
+      giveProcessorsValidFocusTime(OneSignal.sessionManager.getSessionResult(), FocusEventType.BACKGROUND);
+   }
+
+   void onSessionEnded(@NonNull OSSessionManager.SessionResult lastSessionResult) {
+      giveProcessorsValidFocusTime(lastSessionResult, FocusEventType.END_SESSION);
+   }
+
+   void doBlockingBackgroundSyncOfUnsentTime() {
+      if (OneSignal.isForeground())
+         return;
+
+      for(FocusTimeProcessorBase focusTimeProcessor : focusTimeProcessors)
+         focusTimeProcessor.syncUnsentTimeFromSyncJob();
+   }
+
+   private void giveProcessorsValidFocusTime(@NonNull OSSessionManager.SessionResult lastSessionResult, @NonNull FocusEventType focusType) {
+      Long timeElapsed = getTimeFocusedElapsed();
+      if (timeElapsed == null)
+         return;
+
+      for(FocusTimeProcessorBase focusTimeProcessor : focusTimeProcessors)
+         focusTimeProcessor.addTime(timeElapsed, lastSessionResult, focusType);
+   }
+
+   // Get time past since app was put into focus.
+   // Will be null if time is invalid or 0
+   private @Nullable Long getTimeFocusedElapsed() {
+      long timeElapsed = (long)(((SystemClock.elapsedRealtime() - timeFocusedAtMs) / 1_000d) + 0.5d);
+
+      // Time is invalid if below 1 or over a day
+      if (timeElapsed < 1 || timeElapsed > 86_400)
```

Comment:
Maybe we should limit it to the attribution window value (since its changeable) and not just 1 day?

Reply:
This is actually just a sanity validation check before giving it to the `FocusTimeProcessor`.